src/Microsoft.Extensions.Configuration.AzureAppConfiguration/FeatureManagement/FeatureFlagKeyValueAdapter.cs

```diff
@@ -0,0 +1,137 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+//
+using Azure.Data.AppConfiguration;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.Configuration.AzureAppConfiguration.FeatureManagement
+{
+    internal class FeatureFlagKeyValueAdapter : IKeyValueAdapter
+    {
+        private static string _schemaVersion = FeatureManagementConstants.FeatureManagementDefaultSchema;
+
+        public FeatureFlagKeyValueAdapter(string schemaVersion)
+        {
+            if (schemaVersion != FeatureManagementConstants.FeatureManagementSchemaV1
+                && schemaVersion != FeatureManagementConstants.FeatureManagementSchemaV2)
+            {
+                throw new InvalidOperationException(
```

Comment:
As a reader it's unexpected to see InvalidOperationException here instead of ArgumentException. Do we do this because somewhere at a higher layer we handle InvalidOperationException better than ArgumentException? If so, I'd recommend to still throw argumentexception if we throw here, and keep invalidoperationexception as high as possible.

Reply:
That's a good point. We are catching all `InvalidOperationException`s when options are initialized during `Build()` method: https://github.com/Azure/AppConfiguration-DotnetProvider/blob/0ac6b93e7abbb31fbb46090baf244dcc91384447/src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationSource.cs#L55

These exceptions are then ignored if `optional` flag is set to true. So why do we only swallow `InvalidOperationException`s here? Why not all other exceptions thrown during options initialization, like
 `ArgumentException`s? The point of having `optional` set to true was that any errors during AppConfig initialization will be ignored and not crash the app, but here we are ignoring only some exceptions. 

I can see why we might want to throw `ArgumentException` on options initialization - because it helps users quickly fix configuration issues on their end. But by that logic, any exception that occurs during option initialization should be thrown even if `optional` is true. And in some cases, we throw `InvalidOperationException` during options initialization.

Reply:
So it seems the first thing that we need to decide is what the user experience should be like for specifying an invalid version. Which method is going to result in the exception, and what exception should it be.

I would think that it should be thrown when the user calls `UseFeatureFlags()`.

In that case it comes before `Build()` is invoked, and we don't need to worry about optional at that point. It's the same strategy we use for passing an invalid connection string.

Reply:
Not sure I follow. Options are initialized during `Build`: https://github.com/Azure/AppConfiguration-DotnetProvider/blob/0ac6b93e7abbb31fbb46090baf244dcc91384447/src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationSource.cs#L33

Reply:
Oh got it. Thanks, I missed that. Okay, so then we already know wherever an exception, if any, will be thrown.

I see what you mean now. It's a bit inconsistent right now as far as whether an exception will occur when optional is true. In some cases we let the exception bubble up (`ArgumentException`) and some cases we hide it ([example here](https://github.com/Azure/AppConfiguration-DotnetProvider/blob/0ac6b93e7abbb31fbb46090baf244dcc91384447/src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationOptions.cs#L167))

So it sounds like the follow up task is for us to define the behavior we want. Do we want to actually be swallowing argument exceptions for things like an invalid connection string? Or do we want to surface all errors from input parameters like the error I linked above?

My preference would be to fail as early as possible for errors that come from invalid inputs, but that seems to contradict the documentation for the enabled flag.

Reply:
> My preference would be to fail no matter what for errors that come from invalid inputs, but that seems to contradict the documentation for the enabled flag.

My thoughts too. I think our "optional" behavior comes from the "optional" behavior of FileConfigurationProviders. For configuration files, this flag suppresses the exception when a file is not found. Anything else, like invalid syntax in file or empty file path still throws an exception when optional is true. 
Along similar lines, if optional is true for AppConfig provider, any errors that occur during `Build()` should not be suppressed (these are always due to invalid inputs). However, any errors that occur during `Load()` should be suppressed.

cc @zhenlan

Reply:
I think 1. we should do the validation in AzureAppConfiguraionOptions where we read the environment variable 2. and fall back to the default if the environment variable is anything we don't understand (no throw. This will be a good place to add a log). If there is a mismatch between the default schema version and the actual schema, we throw whatever errors later. For most users in the future, they won't set this environment variable, it will be the same experience.